### PR TITLE
OSPF: add support for segment routing extensions as per rfc8665 and r…

### PR DESCRIPTION
**Pull request by hannesgredler**
This is the rebased version of the previous ospf-sr merge request.

### CODE REVIEW

File: print-ospf.c [hunk 1]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
     }
 }
 
+static int
+ospf_print_ri_lsa_sid_label_range_tlv(netdissect_options *ndo, const uint8_t *tptr, u_int tlv_length)
```
Comment (fxlb) [inline print-ospf.c:658]: Should be on 2 lines with indent on second line.

File: print-ospf.c [hunk 2]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+	tptr+=subtlv_length;
+	tlv_length-=subtlv_length;
+    }
+    return 0;
+trunc:
+    return -1;
+}
+
+static int
+ospf_print_ep_lsa_extd_prefix_tlv(netdissect_options *ndo, const uint8_t *tptr, u_int tlv_length)
```
Comment (fxlb) [inline print-ospf.c:711]: Should be on 2 lines with indent on second line.

File: print-ospf.c [hunk 3]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+    while (tlv_length >= 4) {
+
+	ND_TCHECK_4(tptr);
+	subtlv_type = GET_BE_U_2(tptr);
+	subtlv_length = GET_BE_U_2(tptr+2);
+	tptr+=4;
+	tlv_length-=4;
+
+	/* Infinite loop protection. */
+	if (subtlv_type == 0 || subtlv_length == 0) {
```
Comment (fxlb) [inline print-ospf.c:671]: nd_print_invalid(ndo); before the return?

File: print-ospf.c [hunk 4]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+
+    while (tlv_length >= 4) {
+	ND_TCHECK_4(tptr);
+	subtlv_type = GET_BE_U_2(tptr);
+	subtlv_length = GET_BE_U_2(tptr+2);
+	tptr+=4;
+	tlv_length-=4;
+
+	/* Infinite loop protection. */
+	if (subtlv_type == 0 || subtlv_length == 0) {
```
Comment (fxlb) [inline print-ospf.c:724]: nd_print_invalid(ndo); before the return?

File: print-ospf.c [hunk 5]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+    while (lsa_length >= 4) {
+
+	ND_TCHECK_4(tptr);
+	tlv_type = GET_BE_U_2(tptr);
+	tlv_length = GET_BE_U_2(tptr+2);
+	tptr+=4;
+	lsa_length-=4;
+
+	/* Infinite loop protection. */
+	if (tlv_type == 0 || tlv_length == 0) {
```
Comment (fxlb) [inline print-ospf.c:786]: nd_print_invalid(ndo); before the return?

File: print-ospf.c [hunk 6]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+	    }
+
+	    if (prefix_length > 32) {
+		ND_PRINT("\n\t\tIPv4 prefix: bad bit length %u", prefix_length);
+		return -1;
+	    }
+
+	    ND_TCHECK_4(tptr+8);
+	    ND_PRINT("\n\t\tIPv4 prefix: %15s/%u, Route Type: %s, Flags [%s]",
+		     ipaddr_string(ndo, tptr+4), prefix_length,
```
Comment (fxlb) [inline print-ospf.c:815]: Could be GET_IPADDR_STRING().

File: print-ospf.c [hunk 7]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+	    }
+
+	    if (prefix_length > 32) {
+		ND_PRINT("\n\t\tIPv4 prefix: bad bit length %u", prefix_length);
+		return -1;
+	    }
+
+	    ND_TCHECK_4(tptr+8);
+	    ND_PRINT("\n\t\tIPv4 prefix: %15s/%u, Range size: %u, Flags [%s]",
+		     ipaddr_string(ndo, tptr+8), prefix_length,
```
Comment (fxlb) [inline print-ospf.c:846]: Could be GET_IPADDR_STRING().

File: print-ospf.c [hunk 8]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
 }
 
+static int
+ospf_print_ri_lsa_sid_label_range_tlv(netdissect_options *ndo, const uint8_t *tptr, u_int tlv_length)
+{
+    u_int subtlv_type, subtlv_length;
+
+    while (tlv_length >= 4) {
+
+	ND_TCHECK_4(tptr);
```
Comment (fxlb) [inline print-ospf.c:664]: Some ND_TCHECK_*() are useless because there are GET_*() which perform truncation checks.

File: print-ospf.c [hunk 9]
```diff
@@ -836,6 +1099,20 @@ ospf_print_lsa(netdissect_options *ndo,
...
+
+                    case LS_OPAQUE_RI_TLV_HOSTNAME:
+                        ND_PRINT("\n\t      Hostname: ");
+                        nd_printjnp(ndo, tptr, tlv_length);
+                        break;
+
+                    case LS_OPAQUE_RI_TLV_SID_LABEL_RANGE:
+                        ND_TCHECK_4(tptr);
+                        ND_PRINT("\n\t      Range size: %u", GET_BE_U_3(tptr));
+                        if (ospf_print_ri_lsa_sid_label_range_tlv(ndo, tptr+4, tlv_length-4) == -1) {
```
Comment (fxlb) [inline print-ospf.c:1111]: Should be on 2 lines with indent.

**Review by fxlb — CHANGES_REQUESTED** [review]
Some comments.

**Comment by fxlb** [comment]
About the commit message, see CONTRIBUTING.md "Code style and generic remarks" § 10.

**Comment by fxlb** [comment]
You say "This obsoletes #1049", but, if I understand well, some parts are missing, as:
LS_OPAQUE_TYPE_EXTENDED_LINK, LS_OPAQUE_RI_TLV_SR_ALGO, LS_OPAQUE_RI_TLV_SR_LOCAL_BLOCK, LS_OPAQUE_RI_TLV_SRMS_PREFERENCE, ...

File: print-ospf.c [hunk 10]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+
+	default:
+	    if (ndo->ndo_vflag <= 1) {
+		if (!print_unknown_data(ndo, tptr, "\n\t\t", subtlv_length))
+		    return -1;
+	    }
+	}
+
+	/* in OSPF everything has to be 32-bit aligned, including subTLVs */
+	if (subtlv_length % 4) {
```
Comment (fxlb) [inline print-ospf.c:699]: No an invalid case?

**Review by fenner — CHANGES_REQUESTED** [review]
I ran this against the pcap that I included in my old neglected draft pull request.  Other than spacing differences, most of which I'm fine with, this one is missing `SR-Algorithm TLV` parsing.

If you like, I'll create a followon that includes what my patch has that yours doesn't - or you are welcome to pull the test pcap and relevant code from my pull request.

**Comment by fenner** [comment]
> You say "This obsoletes #1049", but, if I understand well, some parts are missing, as: LS_OPAQUE_TYPE_EXTENDED_LINK, LS_OPAQUE_RI_TLV_SR_ALGO, LS_OPAQUE_RI_TLV_SR_LOCAL_BLOCK, LS_OPAQUE_RI_TLV_SRMS_PREFERENCE, ...

I'll create a followon that contains the missing pieces if you'd like to accept Hannes's work here.

**Comment by fxlb** [comment]
> I'll create a followon that contains the missing pieces if you'd like to accept Hannes's work here.

I have no problem with a two-step update, this one and yours.

**Comment by fenner** [comment]
> > I'll create a followon that contains the missing pieces if you'd like to accept Hannes's work here.
> 
> I have no problem with a two-step update, this one and yours.

@hannesgredler does this make sense to you - I'll rebase Arista's changes on yours and create a followon pull request?

File: print-ospf.c [hunk 11]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+	tlv_length = GET_BE_U_2(tptr+2);
+	tptr+=4;
+	lsa_length-=4;
+
+	/* Infinite loop protection. */
+	if (tlv_type == 0 || tlv_length == 0) {
+	    return -1;
+	}
+
+	ND_PRINT("\n\t      %s TLV (%u), length: %u, value: ",
```
[the remark this answers is not in the document]
Comment (fenner) [reply print-ospf.c:790]: What I'm comparing to is, e.g., the RI LSA: https://github.com/the-tcpdump-group/tcpdump/blob/3f975fe6384da20b35401167a19dfbb8f8420408/print-ospf.c#L818

File: print-ospf.c (hunk 1 above)
Comment (hannesgredler) [reply print-ospf.c:658]: ok

File: print-ospf.c (hunk 8 above)
Comment (hannesgredler) [reply print-ospf.c:664]: ok

File: print-ospf.c (hunk 3 above)
Comment (hannesgredler) [reply print-ospf.c:671]: ok

File: print-ospf.c (hunk 10 above)
Comment (hannesgredler) [reply print-ospf.c:699]: no - that's the way it used to be. But admittedly this is a bid messed up in the protocol design ;-)

File: print-ospf.c (hunk 2 above)
Comment (hannesgredler) [reply print-ospf.c:711]: ok

File: print-ospf.c (hunk 4 above)
Comment (hannesgredler) [reply print-ospf.c:724]: ok

File: print-ospf.c (hunk 5 above)
Comment (hannesgredler) [reply print-ospf.c:786]: do you mean I should do this also on the trunc: label ?

File: print-ospf.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (hannesgredler) [reply print-ospf.c:790]: You are correct.

File: print-ospf.c (hunk 6 above)
Comment (hannesgredler) [reply print-ospf.c:815]: ok

File: print-ospf.c (hunk 7 above)
Comment (hannesgredler) [reply print-ospf.c:846]: ok

File: print-ospf.c (hunk 9 above)
Comment (hannesgredler) [reply print-ospf.c:1111]: why is that ? - In line 1111 I am not violating any column boundary ?

**Comment by hannesgredler** [comment]
> > > I'll create a followon that contains the missing pieces if you'd like to accept Hannes's work here.
> > 
> > 
> > I have no problem with a two-step update, this one and yours.
> 
> @hannesgredler does this make sense to you - I'll rebase Arista's changes on yours and create a followon pull request?

Yes certainly - please go ahead.

File: print-ospf.c (hunk 5 above)
Comment (hannesgredler) [reply print-ospf.c:784]: ok

File: print-ospf.c [hunk 12]
```diff
@@ -611,6 +654,226 @@ ospf_print_tos_metrics(netdissect_options *ndo,
...
+
+	    ND_TCHECK_4(tptr+8);
+	    ND_PRINT("\n\t\tIPv4 prefix: %15s/%u, Range size: %u, Flags [%s]",
+		     ipaddr_string(ndo, tptr+8), prefix_length,
+		     range_size,
+		     bittok2str(ep_tlv_flag_values, "none", flags));
+
+	    /* subTLVs present ? */
+	    if (tlv_length > 12) {
+		if (ospf_print_ep_lsa_extd_prefix_tlv(ndo, tptr+12, tlv_length-12) == -1) {
```
Comment (fenner) [inline print-ospf.c:848]: These 12 should be 8 - you subtracted 4 above.

Comment (hannesgredler) [reply print-ospf.c]: You are correct - fixed in latest commit.

File: print-ospf.c (hunk 5 above)
Comment (fxlb) [reply print-ospf.c:783]: `nd_print_invalid(ndo);` still missing.

**Comment by hannesgredler** [comment]
Found three more places where this should get fixed -> https://github.com/the-tcpdump-group/tcpdump/pull/1088/commits/8d53f378d3086c98bd6debb93a0c9968ce96cdfa

**Comment by fxlb** [comment]
Thanks for the updates. Last steps before a merge:

To fix the conflict and have a clean commit:
See `CONTRIBUTING.md` "How to add new code and to update existing code"
`9) Try to rebase your commits to keep the history simple.`
To fix conflict:
$ git rebase upstream/master
To have a clean commit:
$ git rebase -i @~7

To update the commit message:
See  `CONTRIBUTING.md` "Code style and generic remarks"
`10) A commit message must have:`
$ git commit --amend

$ git push -f

**Comment by hannesgredler** [comment]
Rebased and squashed to one commit.

---

commit ac767d55ae6b4e90197b24869fcd0ebcaf6d0808 (HEAD -> master, origin/master, origin/HEAD)
Author: Hannes Gredler ***@***.***>
Date:   Thu Oct 5 15:32:40 2023 +0000

    OSPF: add support for segment routing extensions as per rfc8665 and rfc7684. Add support for hostname extensions as per rfc5642

commit 87c90012f079200b7d49979164e8e9ed89d93d9d (upstream/master)
Author: Francois-Xavier Le Bail ***@***.***>
Date:   Tue Nov 7 11:36:09 2023 +0100

    status exit codes: Remove a no more used enum value

    The old enum value was S_ERR_ND_NO_PRINTER.
    This is a follow-up to 25687e1a312a3bb63a6e8d7f0af0210a615e24a8
    "Print packets for unsupported link-layer protocols in hexadecimal/ASCII".

    [skip ci]
---

**Review by fxlb — APPROVED** [review]

**Comment by fxlb** [comment]
Merged, thank you.
I suggest you to avoid the `master` branch when doing a PR.

**Comment by hannesgredler** [comment]
noted.

On Mon, Nov 13, 2023 at 02:04:08AM -0800, fxlb wrote:
|    Merged, thank you.
|    I suggest you to avoid the master branch when doing a PR.
| 
|    —
|    Reply to this email directly, view it on GitHub, or unsubscribe.
|    You are receiving this because you were mentioned.Message ID:
|    ***@***.***>

**Pull request merged**